runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkInputDataProcessor.java

```diff
@@ -0,0 +1,395 @@
+          createAsync() {
+    return new AsyncSparkInputDataProcessor<>();
+  }
+}
+
+class SyncSparkInputDataProcessor<FnInputT, FnOutputT>
+    implements SparkInputDataProcessor<FnInputT, FnOutputT, Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+  private final UnboundedDoFnOutputManager outputManager = new UnboundedDoFnOutputManager();
+
+  @Override
+  public OutputManager getOutputManager() {
+    return outputManager;
+  }
+
+  @Override
+  public <K> Iterable<Tuple2<TupleTag<?>, WindowedValue<?>>> process(
+      Iterator<WindowedValue<FnInputT>> input, SparkProcessContext<K, FnInputT, FnOutputT> ctx) {
+    if (!input.hasNext()) {
+      return Collections.emptyList();
+    }
+    return () -> new SyncInOutIterator<>(input, ctx);
+  }
+
+  private static class UnboundedDoFnOutputManager
+      implements OutputManager, Iterable<Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+    private final Multimap<TupleTag<?>, WindowedValue<?>> outputs = LinkedListMultimap.create();
+
+    public void clear() {
+      outputs.clear();
+    }
+
+    @Override
+    public Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> iterator() {
+      Iterator<Map.Entry<TupleTag<?>, WindowedValue<?>>> entryIter = outputs.entries().iterator();
+      return Iterators.transform(entryIter, this.entryToTupleFn());
+    }
+
+    private <K, V> Function<Map.Entry<K, V>, Tuple2<K, V>> entryToTupleFn() {
+      return en -> {
+        if (en == null) {
+          return null;
+        } else {
+          return new Tuple2<>(en.getKey(), en.getValue());
+        }
+      };
+    }
+
+    @Override
+    public synchronized <T> void output(TupleTag<T> tag, WindowedValue<T> output) {
+      outputs.put(tag, output);
+    }
+  }
+
+  private class SyncInOutIterator<K>
+      extends AbstractIterator<Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+    private final Iterator<WindowedValue<FnInputT>> inputIterator;
+    private final SparkProcessContext<K, FnInputT, FnOutputT> ctx;
+    private Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> outputIterator;
+    private boolean isBundleStarted;
+    private boolean isBundleFinished;
+
+    SyncInOutIterator(
+        Iterator<WindowedValue<FnInputT>> iterator,
+        SparkProcessContext<K, FnInputT, FnOutputT> ctx) {
+      this.inputIterator = iterator;
+      this.ctx = ctx;
+      this.outputIterator = outputManager.iterator();
+      ;
+    }
+
+    @Override
+    protected Tuple2<TupleTag<?>, WindowedValue<?>> computeNext() {
+      try {
+        // Process each element from the (input) iterator, which produces, zero, one or more
+        // output elements (of type V) in the output iterator. Note that the output
+        // collection (and iterator) is reset between each call to processElement, so the
+        // collection only holds the output values for each call to processElement, rather
+        // than for the whole partition (which would use too much memory).
+        if (!isBundleStarted) {
```

Comment:
Should this check if the bundle is non-empty? I.e. call startBundle only there are elements to be processed?

Reply:
This is just relocation of original code which was there. I did not give it much thought to be honest